src/lib.rs

```diff
@@ -0,0 +1,189 @@
+        PamResultCode::PAM_PERM_DENIED
+    }
+
+    fn sm_setcred(_pamh: &mut PamHandle, _args: Vec<&CStr>, _flags: PamFlag) -> PamResultCode {
+        println!("set credentials");
+        PamResultCode::PAM_SUCCESS
+    }
+
+    fn acct_mgmt(_pamh: &mut PamHandle, _args: Vec<&CStr>, _flags: PamFlag) -> PamResultCode {
+        println!("account management");
+        PamResultCode::PAM_SUCCESS
+    }
+}
+
+async fn generate_new_token(state: &str, port: u16) -> String {
+    let addr = ([0, 0, 0, 0], port).into();
+
+    let notify = Arc::new(Notify::new());
+    let token = Arc::new(Mutex::new(None::<String>));
+    let make_svc = make_service_fn({
+        let notify = notify.clone();
+        let token = token.clone();
+        move |_| {
+            let notify = notify.clone();
+            let token = token.clone();
+            async move {
+                Ok::<_, Infallible>(service_fn(move |req: Request<Body>| {
+                    let notify = notify.clone();
+                    let token = token.clone();
+                    async move {
+                        let code = req
+                          .uri()
+                          .query()
+                          .map(|query|
+                            url::form_urlencoded::parse(query.as_bytes())
+                          )
+                          .and_then(|mut query|
+                            query.find(|(name, _)| name.to_string() == "code")
+                          )
+                          .map(|(_, code)| code.into_owned());
+                        if let (Some(code), token @ None) = (code, &mut *token.lock().await) {
+                          notify.notify_one();
+                          *token = Some(code);
+                        }
+                        Ok::<_, Infallible>(Response::new(Body::from(
+                            "fuck you",
+                        )))
+                    }
+                }))
+            }
+        }
+    });
+    let server = hyper::Server::bind(&addr)
+        .serve(make_svc)
+        .with_graceful_shutdown(async move { tokio::select! {
+          _ = notify.notified() => (),
+          _ = sleep(Duration::from_secs(60 * 1)) => (),
+        }});
+    server.await.unwrap();
+
+    let mut lock = token.lock().await;
+    lock.take().unwrap()
+}
+
+async fn get_access_token(client_id: &str, client_secret: &str, code: String, url: String) -> Result<std::string::String, reqwest::Error> {
+    let data = [("client_id", client_id), ("client_secret", client_secret), ("grant_type", "authorization_code"), ("code", &code), ("redirect_uri", &url)];
+    let client = reqwest::Client::new();
+    let res = client.post(format!("{}/v10/oauth2/token", API_ENDPOINT))
+            .header("Content-Type", "application/x-www-form-urlencoded")
+            .form(&data)
+            .send()
+            .await?;
+    return res.text().await;
+}
```

Comment:
For all of these API requests, we're getting JSON back. You might wanna use the `res.json()` function and define structs that represent the JSON structure you expect to get back. That way you get to write less parse/deserializing code yourself, plus you can nicer errors!